File: app/src/main/java/moe/frisk/myhooks/util/AndroidUi.java
package moe.frisk.myhooks.util;

import android.app.AlertDialog;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.widget.Toast;

public final class AndroidUi {

    private static final Handler MAIN_HANDLER = new Handler(Looper.getMainLooper());

    private AndroidUi() {
    }

    public static void runOnMainThread(Runnable block) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            block.run();
        } else {
            MAIN_HANDLER.post(block);
        }
    }

    public static void toast(final Context context, final String text) {
        runOnMainThread(new Runnable() {
            @Override
            public void run() {
                Toast.makeText(context, text, Toast.LENGTH_SHORT).show();
            }
        });
    }

    public static void copyText(Context context, String label, String text) {
        ClipboardManager clipboard =
            (ClipboardManager) context.getSystemService(Context.CLIPBOARD_SERVICE);
        clipboard.setPrimaryClip(ClipData.newPlainText(label, text));
    }

    public static void showDialog(Context context, String title, String message) {
        showDialog(context, title, message, "关闭", null, null, null, null, null);
    }

    public static void showDialog(final Context context,
                                  final String title,
                                  final String message,
                                  final String positiveText,
                                  final Runnable onPositive,
                                  final String negativeText,
                                  final Runnable onNegative,
                                  final String neutralText,
                                  final Runnable onNeutral) {
        runOnMainThread(new Runnable() {
            @Override
            public void run() {
                AlertDialog.Builder builder = new AlertDialog.Builder(context)
                    .setTitle(title)
                    .setMessage(message)
                    .setPositiveButton(positiveText, new android.content.DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(android.content.DialogInterface dialog, int which) {
                            if (onPositive != null) {
                                onPositive.run();
                            }
                        }
                    });
                if (negativeText != null) {
                    builder.setNegativeButton(negativeText, new android.content.DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(android.content.DialogInterface dialog, int which) {
                            if (onNegative != null) {
                                onNegative.run();
                            }
                        }
                    });
                }
                if (neutralText != null) {
                    builder.setNeutralButton(neutralText, new android.content.DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(android.content.DialogInterface dialog, int which) {
                            if (onNeutral != null) {
                                onNeutral.run();
                            }
                        }
                    });
                }
                builder.show();
            }
        });
    }
}

File: app/src/main/java/moe/frisk/myhooks/util/AndroidUi.kt
package moe.frisk.myhooks.util

import android.app.AlertDialog
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.os.Handler
import android.os.Looper
import android.widget.Toast

object AndroidUi {

    private val mainHandler = Handler(Looper.getMainLooper())

    fun runOnMainThread(block: () -> Unit) {
        if (Looper.myLooper() == Looper.getMainLooper()) {
            block()
        } else {
            mainHandler.post(block)
        }
    }

    fun toast(context: Context, text: String) {
        runOnMainThread {
            Toast.makeText(context, text, Toast.LENGTH_SHORT).show()
        }
    }

    fun copyText(context: Context, label: String, text: String) {
        val clipboard = context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        clipboard.setPrimaryClip(ClipData.newPlainText(label, text))
    }

    fun showDialog(
        context: Context,
        title: String,
        message: String,
        positiveText: String = "关闭",
        onPositive: (() -> Unit)? = null,
        negativeText: String? = null,
        onNegative: (() -> Unit)? = null,
        neutralText: String? = null,
        onNeutral: (() -> Unit)? = null,
    ) {
        runOnMainThread {
            val builder = AlertDialog.Builder(context)
                .setTitle(title)
                .setMessage(message)
                .setPositiveButton(positiveText) { _, _ -> onPositive?.invoke() }
            if (negativeText != null) {
                builder.setNegativeButton(negativeText) { _, _ -> onNegative?.invoke() }
            }
            if (neutralText != null) {
                builder.setNeutralButton(neutralText) { _, _ -> onNeutral?.invoke() }
            }
            builder.show()
        }
    }
}
